### PR TITLE
docs: add docs for enabling automated TLS and DNS for v3 on GCP

### PR DESCRIPTION
**Pull request by rawlingsj**

### CODE REVIEW

**Comment by rawlingsj** [comment]
lets validate before merging
/hold

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:14]: I think the correct file name is jx-requirements.yml.

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:38]:
```suggestion
Most people prefer to use a subdomain for a specific installation rather than purchasing one domain per cluster.  For example in a multi cluster setup you will probably want the same parent domain but each cluster using a different subdomain like development.foo.io, staging.foo.io and foo.io.
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:67]:
```suggestion
- Install [cert-manager](https://cert-manager.io/docs/) - Kubernetes controller which watches for requests to ask [Let's Encrypt](https://letsencrypt.org/) to issue a new wildcard TLS certificate for your domain and will manage this including renewals
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:107]:
```suggestion
The domain from setting up your infrastructure in step one should appear in the `jx-requirements.yml` of you cluster git repo.  Next configure your TLS options, update your `jx-requirements.yml` with
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:20]:
```suggestion
## Prerequisites
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [reply content/en/docs/v3/guides/infra/tls_dns.md:38]: `foo.io` is a parent domain not a subdomain

File: content/en/docs/v3/guides/infra/google_cloud_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/google_cloud_dns.md:17]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/google_cloud_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/google_cloud_dns.md:25]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:48]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:54]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:73]:
```suggestion
```yaml
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:78]:
```suggestion
```yaml
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:86]:
```suggestion
```yaml
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:98]:
```suggestion
```yaml
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:108]:
```suggestion
```yaml
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:123]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:127]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:134]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/upgrade.md
Comment (borntorock) [inline content/en/docs/v3/guides/upgrade.md:27]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/upgrade.md
Comment (borntorock) [inline content/en/docs/v3/guides/upgrade.md:38]:
```suggestion
```bash
```

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/tls_dns.md:10]:
```suggestion
To achieve this we will use a couple of open source projects to help enable automated DNS for your applications.
```

File: content/en/docs/v3/guides/infra/cloud-config.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/cloud-config.md:20]:
```suggestion
* <a href="https://github.com/jenkins-x/terraform-google-jx#inputs" class="btn bg-primary text-light">GCP</a>

* <a href="https://github.com/jenkins-x/terraform-aws-eks-jx#inputs" class="btn bg-primary text-light">AWS</a>

* <a href="https://github.com/jenkins-x/terraform-azurerm-jx#inputs" class="btn bg-primary text-light">Azure</a>
```

File: content/en/docs/v3/guides/infra/cloud-config.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/cloud-config.md:18]:
```suggestion
* <a href="https://github.com/jenkins-x/terraform-aws-eks-jx#inputs" class="btn bg-primary text-light">AWS</a>
```

File: content/en/docs/v3/guides/infra/cloud-config.md
Comment (borntorock) [inline content/en/docs/v3/guides/infra/cloud-config.md:16]:
```suggestion
* <a href="https://github.com/jenkins-x/terraform-google-jx#inputs" class="btn bg-primary text-light">GCP</a>
```

File: content/en/docs/v3/guides/infra/cloud-config.md
Comment (ankitm123) [inline content/en/docs/v3/guides/infra/cloud-config.md:16]: Better to use hugo shortcode insteaf of html in markdown.

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (chrismellard) [inline content/en/docs/v3/guides/infra/tls_dns.md:89]: might want to add something like this to highlight the dependency on cert-manager so we avoid race conditions - chart: jetstack/cert-manager
  name: cert-manager
  namespace: jx
- chart: jx3/acme
  needs:
    - cert-manager
  name: tls-jx
  namespace: jx
  values:
    - issuer:
        enabled: true
        cluster: true

**Review by chrismellard — CHANGES_REQUESTED** [review]

File: content/en/docs/v3/guides/infra/tls_dns.md
Comment (rawlingsj) [reply content/en/docs/v3/guides/infra/tls_dns.md]: thanks!  we might need to think about this one a little further because Jenkins X uses `helmfile template` with an output directory to write the generated kubernetes resources, it then does `kubectl apply` so I'm not totally sure if the 
```
needs:
  - cert-manager/cert-manager
```
will work as expected, there is a an option helmfile template flag
```
--output-dir-template value   go text template for generating the output directory. Default: {{ .OutputDir }}/{{ .State.BaseName }}-{{ .State.AbsPathSHA1 }}-{{ .Release.Name}}
```
which might be useful to see if we can split the CRDs into a different folder from the `ClusterIssuer` and then `kubectl` applies them in the right order?

i.e. 
```
helmfile template --output-dir=cluster --output-dir-template=[something]
```
generates folders:
```
/cluster
- /crds/foo.yaml
- /other cluster resources/bar.yaml
```
then 
```
kubectl apply -f cluster
```
first applies the CRDs and next the other cluster resources?

Comment (chrismellard) [reply content/en/docs/v3/guides/infra/tls_dns.md]: Think you're right! I think the two options are the one  you have suggested where we parse CRDs and modify their path so they are applied first. The only other one which feels like a worse  option is we support nested helmfiles and honour this structure in our `cluster` and `namespace` folders. I can't really think of much of a use case for  nested state  unless our chart list gets very large so for now probably your suggestion of separating and applying  CRDs before other resources makes the most sense. Sounds like an issue against `jx gitops` which I'll open later

**Pull request merged**
